src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+        }
+
+        if (vectors != null && !vectors.isEmpty()) {
+            queryRequest.addAllVector(vectors);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
+            if (sparseIndices.size() != sparseValues.size()) {
+                throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+            }
+
+            queryRequest.setSparseVector(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        return blockingStub.query(queryRequest.build());
+    }
+
+    public FetchResponse fetch(List<String> ids,
+                               String namespace) {
+        FetchRequest.Builder fetchRequest = FetchRequest.newBuilder();
+
+        if (ids == null || ids.isEmpty()) {
+            throw new PineconeValidationException("Invalid Fetch Request. Vector IDs must be present");
+        }
+
+        fetchRequest.addAllIds(ids);
+
+        if (namespace != null && !namespace.isEmpty()) {
+            fetchRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.fetch(fetchRequest.build());
+    }
+
+    public UpdateResponse update(String id,
+                                 List<Float> values,
+                                 Struct metadata,
+                                 String namespace,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues) {
+        UpdateRequest.Builder updateRequest = UpdateRequest.newBuilder();
+
+        if (id == null) {
+            throw new PineconeValidationException("Invalid Update Request. Vector ID must be present");
+        }
+        updateRequest.setId(id);
+
+        if (values != null) {
+            updateRequest.addAllValues(values);
+        }
+
+        if (metadata != null) {
+            updateRequest.setSetMetadata(metadata);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            updateRequest.setNamespace(namespace);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
+            if (sparseIndices.size() != sparseValues.size()) {
+                throw new PineconeValidationException("Invalid Update Request. Please ensure that both sparse indices and values are of the same length.");
+            }
+
+            updateRequest.setSparseValues(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        return blockingStub.update(updateRequest.build());
+    }
+
+    public DeleteResponse delete(List<String> ids,
+                                 boolean deleteAll,
+                                 String namespace, Struct filter) {
```

Comment:
Since most of the arguments don't make sense when used together, what do you think about this method breakdown:

- `delete(List<String> ids, Struct filter, boolean deleteAll, String namespace)`
- `deleteByIds(List<String> ids, String namespace)`
- `deleteByIds(List<String> ids)` for default namespace
- `deleteByFilter(Struct filter, String namespace)`
- `deleteByFilter(Struct filter)` for default namespace
- `deleteAll(String namespace)`

And as an additional convenience:
- `deleteById(String id, String namespace)` for one single id
- `deleteById(String id)` for one single id in default namespace

Thoughts @austin-denoble @haruska?

Maybe I'm misreading the API docs, but `deleteAll` is about clearing out vectors in a namespace which seems incompatible with specifying ids/filters, so that should also be checked as an error case.

Reply:
Added different options